operators/tokenizer/bpe_utils.hpp

```diff
@@ -198,353 +299,435 @@ class TokenWithRegularExp {
+            category_set.categories.push_back(single_category_match.str(1));
+        }
 
-    // \s+(?!\S)|\s+
-    if ((m_text.size() >= 1) && (IsZ(m_text[0]))) {
-      size_t i = 1;
-      for (; i < m_text.size(); ++i) {
-        if (!IsZ(m_text[i])) break;
+        category_sets.push_back(category_set);
       }
-      if ((i > 1) && (i != m_text.size())) {  //\s+(?!\S)
-        i--;
-        std::u32string_view res = m_text.substr(0, i);
-        m_text = m_text.substr(i);
-        return res;
+
+      std::u32string m_text_copy = m_text;
+      std::u32string concatenated = U"";
+      bool skip = false;
+
+      // Perform RegexCategory on each category set
+      for (const auto& set : category_sets) {
+          if (!skip) {
+            std::u32string res = RegexCategory(set.categories, set.premodifier, set.postmodifier, set.negated);
+            if (res != U""){
+              concatenated.append(res);
+            } else if (set.postmodifier == "+" && (set.categories.size() == 1 ? (set.categories[0] != "rn") : true)){
+              // + requires a match of one or more, so if there is not at least one, the whole set has no match
+              m_text = m_text_copy;
+              skip = true;
+            }
+          }
       }
-      // \s+
-      std::u32string_view res = m_text.substr(0, i);
-      m_text = m_text.substr(i);
-      return res;
+
+      if (concatenated != U"" && !skip)
+        return concatenated;
     }
 
-    return std::u32string_view{};
+    return U"";
   }
 
   // RegexMatchCustom takes a regular expression 'regex_expr' and perform pattern matching to get the next token.
-  // If the regex is familiar and matches that for one of our pre-written parsers for GPT2 or Llama3, we use the
-  // appropriate methods. If not, we check to see if the regex can be parsed with the C++ Standard Library methods.
-  std::u32string_view RegexMatchCustom(const std::string & regex_expr) {
-    std::vector<size_t> bpe_offsets;
-
-    if (regex_expr == GPT2_REGEX_PATTERN) {
-        return RegexMatchGPT2();
-    } else if (regex_expr == LLAMA_REGEX_PATTERN_1 ||
-               regex_expr == LLAMA_REGEX_PATTERN_2) {
-
-        return RegexMatchLlama3();
-    }
-
+  // If the regex can be parsed by our general RegexMatchGeneral parser designed to handle the majority of regex cases
+  // it will be taken care of there. If not, we check to see if the regex can be parsed with the C++ Standard Library methods.
+  std::u32string RegexMatchCustom(const std::string & regex_expr) {
     try {
-      return RegexMatchSTD(ustring(regex_expr));
+      std::u32string res = RegexMatchGeneral(regex_expr);
+      if (res != U"")
+        return res;
+      else
+        return RegexMatchSTD(ustring(regex_expr));
     } catch (const std::exception& /* ex */) {
       std::string part1 = "Regex '";
       std::string part2 = "' not supported!";
       throw std::runtime_error(part1 + regex_expr + part2);
```

Comment:
can you replace all of throw with ORTX_CXX_API_THROW?

Reply:
done!